src/Engine/Graphics/Collisions.cpp

```diff
@@ -749,129 +872,147 @@ void ProcessPartyCollisionsBLV(int sectorId, int min_party_move_delta_sqr, int *
-        int new_party_z_tmp = pParty->pos.z +
-                              collision_state.adjusted_move_distance * collision_state.direction.z;
+        pParty->pos = adjusted_pos;
 
         if (collision_state.pid.type() == OBJECT_Actor) {
             if (pParty->pPartyBuffs[PARTY_BUFF_INVISIBILITY].Active())
                 pParty->pPartyBuffs[PARTY_BUFF_INVISIBILITY].Reset(); // Break invisibility when running into a monster.
         }
 
         if (collision_state.pid.type() == OBJECT_Decoration) {
-            // Bounce back from a decoration & do another round of collision checks.
-            // This way the party can "slide" along & past a decoration.
-            int angle = TrigLUT.atan2(pParty->pos.x - pLevelDecorations[collision_state.pid.id()].vPosition.x,
-                                      pParty->pos.y - pLevelDecorations[collision_state.pid.id()].vPosition.y);
-            int len = integer_sqrt(pParty->speed.xy().lengthSqr());
-            pParty->speed.x = TrigLUT.cos(angle) * len;
-            pParty->speed.y = TrigLUT.sin(angle) * len;
+            // TODO(pskelton): common to odm/blv so extract
+            Vec3f newdirection;
+            if (collision_state.adjusted_move_distance > 0.0f) {
+                // Create new sliding plane from collisio
+                Vec3f slideplaneorigin = collision_state.collisionPos;
+                Vec3f slideplanenormal = collision_state.collisionNorm;
+                float slideplanedist = -(dot(slideplaneorigin, slideplanenormal));
+
+                // Form a sliding vector that is parallel to sliding movement
+                // Take where you wouldve ended up without collisions and move that onto the slide plane by adding the normal
+                // Start point to new destination is a vector along the slide plane
+                float destplanedist = dot(collision_state.new_position_lo, slideplanenormal) + slideplanedist;
+                Vec3f newdestination = collision_state.new_position_lo - destplanedist * slideplanenormal;
+                newdirection = newdestination - collision_state.collisionPos;
+                newdirection.z = 0;
+                newdirection.normalize();
+            }
+
+            // Set party to move along this new sliding vector
+            pParty->speed = newdirection * dot(newdirection, pParty->speed);
+            // Skip reducing party speed
+            continue;
         }
 
         if (collision_state.pid.type() == OBJECT_Face) {
             BLVFace *pFace = &pIndoor->pFaces[collision_state.pid.id()];
+            bool bFaceSlopeTooSteep = pFace->facePlane.normal.z > 0.0f && pFace->facePlane.normal.z < 0.70767211914f; // Was 46378 fixpoint
+
+            // TODO(pskelton): Better way to do this?
+            // Special case for steep staircase in tidewater
+            if (pCurrentMapName == "D17.blv") {
+                if (collision_state.pid.id() == 650)
+                    bFaceSlopeTooSteep = false;
```

Comment:
Oh lol, how did it work before? X)

Reply:
Previously you could carry a bit of upwards momentum into the next frame that allowed you to overcome the slope. You could occasionally get a bit of air at the top of stairs.

Reply:
Let's add a TODO. We'll need a better fix for this.